src/lib/openssl_preload/crypto.c

```diff
@@ -0,0 +1,232 @@
+/*
+ * The Shadow Simulator
+ * Copyright (c) 2010-2011, Rob Jansen
+ * See LICENSE for licensing information
+ */
+
+#define _GNU_SOURCE
+#include <assert.h>
+#include <dlfcn.h>
+#include <execinfo.h>
+#include <pthread.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#ifdef DEBUG
+#define debuglog(...) fprintf(stderr, __VA_ARGS__)
+#else
+#define debuglog(...)
+#endif
+
+// Caches whether or not a memory address is from libssl.so
+typedef struct {
+    void* addr;
+    bool is_libssl;
+} bt_cache_entry_t;
+
+// Lock for safely accessing this lib's global state from multiple threads.
+// TODO: We currently do no error checking when operating on this lock.
+pthread_mutex_t global_state_lock = PTHREAD_MUTEX_INITIALIZER;
+
+// We use a small cache size: in ad-hoc experiments with tor-0.4.6.9, we observed
+// at most three callers of EVP_EncryptUpdate.
+#define EVP_BACKTRACE_CACHE_LEN 10
+static bt_cache_entry_t evp_backtrace_cache[EVP_BACKTRACE_CACHE_LEN];
+
+// For storing a pointer to the original EVP_EncryptUpdate function.
+typedef int EVP_EncryptUpdate_func(void*, unsigned char*, int*, const unsigned char*, int);
+static void* evp_eu_funcptr = NULL;
+
+// Counters for verifying that interception is happening correctly.
+static unsigned long aes_e_cnt = 0;
+static unsigned long aes_d_cnt = 0;
+static unsigned long aes_ce_cnt = 0;
+static unsigned long crypto_ce_cnt = 0;
+static unsigned long crypto_cec_cnt = 0;
+static unsigned long evp_c_cnt = 0;
+static unsigned long evp_eu_cnt = 0;
+
+static void _print_counters() {
+    debuglog("Counters: {'AES_encrypt':%lu, 'AES_decrypt':%lu, 'AES_ctr128_encrypt':%lu, "
+             "'CRYPTO_ctr128_encrypt':%lu, 'CRYPTO_ctr128_encrypt_ctr32':%lu, "
+             "'EVP_Cipher':%lu, 'EVP_EncryptUpdate':%lu}\n",
+             aes_e_cnt, aes_d_cnt, aes_ce_cnt, crypto_ce_cnt, crypto_cec_cnt, evp_c_cnt,
+             evp_eu_cnt);
+}
+
+__attribute__((constructor)) void _crypto_load() {
+    debuglog("Loading the preloaded crypto interception lib\n");
+
+    int result = pthread_mutex_lock(&global_state_lock);
+    assert(result == 0);
```

Comment:
asserts are disabled in debug builds.

I think we want something like `if (result != 0) { abort(); }`.

Or if we want to pull in the logging library here (I don't think there's a reason not to...)  `if (result != 0) { panic("pthread_mutex_lock: %s", strerror(result));}`

Reply:
Done.